### DISCUSSION TITLE
Nesting a read within a write

### DISCUSSION THREAD

**Discussion by jamesonev** [prisma/prisma#3569] [Q&A]
I'm trying to use Prisma to create a new tuple that is a modification of another tuple and I need it to happen in a single transaction, but I keep getting stuck. I'd rather now use `$queryRaw` or `$executeRaw`. Is there an elegant way to make a write depend on the contents of a read?

I'm using prisma `2.6.2`

**Comment by ryands17** [comment]
Hey @jamesonev 👋 
Could you be a bit more specifc in terms of what you're trying to achieve? Maybe share your `schema.prisma` and the code with Prisma queries.

**Comment by jamesonev** [comment]
We're using Prisma as our DB interface for a double-entry accounting system, and so every "update" is really a read of the most recent values and then an insert of those values but modified. Here's an example. Assume every user starts with 500.

|id|amount|debitFrom|creditTo|debitorNewBalance|creditorNewBalance|
|-|-|-|-|-|-|
|1|200|Alice|Bob|300|700|
|2|50|Cole|Alice|450|350|

So now to create a new transaction (let's say from Bob to Cole), we need to read each of their most recent balances and then create a new row, so our table will look like this:

|id|amount|debitFrom|creditTo|debitorNewBalance|creditorNewBalance|
|-|-|-|-|-|-|
|1|200|Alice|Bob|300|700|
|2|50|Cole|Alice|450|350|
|3|100|Bob|Cole|600|550|

For the new balances to be correct, I have to guarantee that nothing happens between the reads of the balances and the creation of the new row. I think the query should be something like this: 
```ts
prisma.balance.create({
  data: {
    amount: 100,
    debitFrom: { connect: { name: from.name } },
    creditTo: { connect: { name: to.name } },
    debitorNewBalance: { [FIND BOB'S LAST BALANCE] - amount },
    creditorNewBalance: { [FIND COLE'S LAST BALANCE] + amount },
  },
})
```
and we could find Bob's last balance using something like this
```ts
prisma.balance.findMany({
  where: { debitFrom: Bob OR creditTo: Bob },
  take: 1,
  orderBy: 'desc',
})[0] // only take the first result
```
But I'm unsure of how to actually put this balance.findMany query into my balance.create query.

**Reply by ryands17** [reply]
@jamesonev 

I think you could perform this in two queries:

```ts
const [bobsBalance, colesBalance] = await Promise.all([
prisma.balance.findMany({
  where: { debitFrom: Bob },
  take: 1,
  orderBy: 'desc',
}),
prisma.balance.findMany({
  where: { debitFrom: Cole },
  take: 1,
  orderBy: 'desc',
}),
])

// and then perform your original request

prisma.balance.create({
  data: {
    amount: 100,
    debitFrom: { connect: { name: from.name } },
    creditTo: { connect: { name: to.name } },
    debitorNewBalance: bobsBalance - amount ,
    creditorNewBalance: colesBalance + amount,
  },
})
```

Let me know if that helps.

**Reply by jamesonev** [reply]
@ryands17 thanks for thinking through this with me. This read then write strategy is what we're doing right now, but it's not sufficient because balances can change between the read and write (it's a classic race condition). I'm wondering if it's possible to create a single transaction with all 3 of these queries, or if there's some way to nest them so that we get ACID properties. 

```
const [bobsBalance, colesBalance] = await Promise.all([
prisma.balance.findMany({
  where: { debitFrom: Bob },
  take: 1,
  orderBy: 'desc',
}),
prisma.balance.findMany({
  where: { debitFrom: Cole },
  take: 1,
  orderBy: 'desc',
}),
])

// bob's balance could be modified here by some external request, and then that change would be overwritten by the insert below

// and then perform your original request

prisma.balance.create({
  data: {
    amount: 100,
    debitFrom: { connect: { name: from.name } },
    creditTo: { connect: { name: to.name } },
    debitorNewBalance: bobsBalance - amount ,
    creditorNewBalance: colesBalance + amount,
  },
})
```
